src/Language/PureScript/Sugar/BindingGroups.hs

```diff
@@ -103,9 +104,17 @@ createBindingGroups moduleName = mapM f <=< handleDecls
           in (d, (name, vty), self ++ deps)
         dataVerts = fmap mkVert allDecls
     dataBindingGroupDecls <- parU (stronglyConnCompR dataVerts) toDataBindingGroup
-    let allIdents = fmap valdeclIdent values
-        valueVerts = fmap (\d -> (d, valdeclIdent d, usedIdents moduleName d `intersect` allIdents)) values
-    bindingGroupDecls <- parU (stronglyConnComp valueVerts) (toBindingGroup moduleName)
+    let 
+      makeValueDeclarationKey = (,) <$> exprHasNoTypeHole . valdeclExpression <*> valdeclIdent
+      valueDeclarationKeys = makeValueDeclarationKey <$> values
+
+      valueDeclarationInfo = M.fromList $ swap <$> valueDeclarationKeys
+      computeValueDependencies = (`intersect` valueDeclarationKeys) . usedIdents valueDeclarationInfo moduleName 
+  
+      makeValueDeclarationVert = (,,) <$> id <*> makeValueDeclarationKey <*> computeValueDependencies
+      valueDeclarationVerts = makeValueDeclarationVert <$> values
+
+    bindingGroupDecls <- parU (stronglyConnComp valueDeclarationVerts) (toBindingGroup moduleName)
```

Comment:
Am I correct that the big change here is that you've changed the keys of the dependency graph from `Ident` to `(Bool, Ident)` and you're relying on the implicit lexicographic sorting of `stronglyConnComp` to put hole-less declarations first?

That could be a little more clear, if so. I don't see anything in the documentation for `Data.Graph` that directly claims this as an invariant.

Reply:
That's correct. I could also probably use a custom data type for this if we ever want to do reordering beyond hole or no-hole declarations.

After having asked, it seems like this is more of an implementation detail rather than an invariant. That being said, implementing a topological sort that maintains this invariant is definitely possible.